### PR TITLE
Feat/delete api

### PR DESCRIPTION
**Pull request by zenovore**
Adding Deletion Feature on Turing (For ensembler and ensembling jobs), change update feature to cleanup previous mlflow run

### Delete Feature
This PR also changes the UI of Turing by adding a delete button on the ensembler and ensembling job pages. Deleting an ensembler has several constraints that must be met, including:
- No `deployed` or `pending` router versions that use the ensembler.
- No `pending`, `building`, or `running` ensembling jobs that use the ensembler.

If the ensembler can be deleted, this feature will also delete other related entities, such as:
- Router version with `failed` / `undeployed` status
- Ensembling job with `failed` / `completed` / `failed_building` / `failed_submission` status

A popup will appear showing the entity that is blocking the deletion process (if the constraints have not been met) and the entities that will be deleted if the constraints are met.

### Update Feature
We have added a check on the SDK that blocks the update process if certain conditions exist, including:
- Router versions with a `pending` status.
- Ensembling jobs with `building`, `running`, or `pending` status.

In addition, we have added a feature to delete old MLflow artifacts and runs when the update is successful. This means that the old MLflow run, including the artifact, will also be deleted

### To Do List 
- [x] Deletion API
- [x] Deletion SDK
- [x] Deletion Frontend
- [x] API Unit Test

### CODE REVIEW

File: api/turing/api/router_versions_api.go [hunk 1]
```diff
@@ -177,6 +180,24 @@ func (c RouterVersionsController) DeployRouterVersion(
 		"version":   int(routerVersion.Version),
 	})
 }
+func (c RouterVersionsController) ListEnsemblerRouterVersion(
```
Comment (deadlycoconuts) [inline api/turing/api/router_versions_api.go:183]: Hmm this name doesn't sound too descriptive about what it does, can we rename it to something like `ListRouterVersionsByEnsembler` or `ListRouterVersionsWithEnsembler`?

File: sdk/turing/batch/job.py [hunk 2]
```diff
@@ -145,6 +145,7 @@ def list(
         status: List[EnsemblingJobStatus] = None,
         page: Optional[int] = None,
         page_size: Optional[int] = None,
+        ensembler_id: Optional[int] = None,
```
Comment (deadlycoconuts) [inline sdk/turing/batch/job.py:148]: Why is this added to the signature? Does this even work? If I remembered correctly, the API `ListEnsemblingJobs` endpoint does not support the `ensembler_id` as a query param right?

File: api/turing/api/ensemblers_api.go [hunk 3]
```diff
@@ -109,18 +111,162 @@ func (c EnsemblersController) UpdateEnsembler(
...
+
 	if err = ensembler.Patch(request.EnsemblerLike); err != nil {
 		return BadRequest("invalid ensembler configuration", err.Error())
 	}
-
 	ensembler, err = c.EnsemblersService.Save(ensembler)
 	if err != nil {
+		if isPyfunc {
+			_ = c.MlflowService.DeleteRun(context.Background(), pyFuncEnsembler.RunID, pyFuncEnsembler.ArtifactURI, true)
+		}
```
[the remark this answers is not in the document]
Comment (zenovore) [reply api/turing/api/ensemblers_api.go:131]: yep, that's correct, if the update failed, then it will delete the new mlflow run, the reason i dont handle this error were to avoid confusion, e.g user update, then get failed message that this error happens because we couldn't remove the new run

File: api/turing/api/ensemblers_api.go [hunk 4]
```diff
@@ -109,18 +111,162 @@ func (c EnsemblersController) UpdateEnsembler(
...
 		return InternalServerError("failed to update an ensembler", err.Error())
 	}
 
+	// Delete If Only RunID is changed, of only name is changed, doesn't need to delete the ensembler run
+	if isPyfunc {
+		_ = c.MlflowService.DeleteRun(context.Background(), oldPyFuncEnsembler.RunID, oldPyFuncEnsembler.ArtifactURI, true)
```
[the remark this answers is not in the document]
Comment (zenovore) [reply api/turing/api/ensemblers_api.go:137]: This were the deletion of the old mlflow run, if the update already succeed, we need to delete the old mlflow run, and if the deletion were failed, to avoid confusion we just let it be, since if we show an error message, but the update were succeed, it will confuse the user

File: sdk/turing/batch/job.py (hunk 2 above)
Comment (zenovore) [reply sdk/turing/batch/job.py:148]: Yes this is working, https://github.com/caraml-dev/turing/blob/58f66852ccb6861cda4a4854d6e658ebd72cf8d3/api/turing/service/ensembling_job_service.go#L178-L180
we could search list of jobs by ensembler 😁

File: api/turing/api/ensemblers_api.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (krithika369) [reply api/turing/api/ensemblers_api.go:138]: Hmm.. We discussed something similar with the mlflow run deletion earlier, around the deletion of the GCS artifact. I'm of the opinion that the nested artifact should be deleted first and errors should block further processing. So in this case, can we delete the old run first before saving the new one in the Turing DB? If deletion of the old run failed, we can just show error to the user. Wdyt?

**Review by krithika369 — COMMENTED** [review]
Still in the process of reviewing. Left some comments first. Can we also include the user docs for the UI changes in this PR ?

File: api/turing/api/ensemblers_api.go [hunk 5]
```diff
@@ -109,18 +111,160 @@ func (c EnsemblersController) UpdateEnsembler(
...
+	}
+
 	if err = ensembler.Patch(request.EnsemblerLike); err != nil {
 		return BadRequest("invalid ensembler configuration", err.Error())
 	}
-
 	ensembler, err = c.EnsemblersService.Save(ensembler)
 	if err != nil {
+		// Delete If Only RunID is changed, if only name is changed, doesn't need to delete the mlflow run
+		if isPyfunc && oldPyFuncEnsembler.RunID != pyFuncEnsembler.RunID {
```
[the remark this answers is not in the document]
Comment (zenovore) [reply api/turing/api/ensemblers_api.go:130]: If the only thing that updated were the names, mlflow wont create another run, so we don't need to delete old and new run, we just need to change the database

> Why are we deleting the old runs if there is an error?

When he update process failed, we delete the new mlflow run, since if we update the ensembler from the sdk, the sdk already create a new mlflow run, so when the update is failed, the new mlflow run will get deleted.

File: api/turing/api/ensemblers_api.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (zenovore) [reply api/turing/api/ensemblers_api.go:138]: if we delete the old mlflow run before updating the db, and when the db update failed, then the ensembler can't be used again since the mlflow run + gcs artifact is already deleted, that's the main reason why i put the deletion of the old ensembler last

File: api/api/specs/routers.yaml [hunk 6]
```diff
@@ -420,6 +420,38 @@ paths:
         404:
           description: "No router version found"
 
+  "/projects/{project_id}/routers-version-ensembler/{ensembler_id}":
```
[the remark this answers is not in the document]
Comment (zenovore) [reply api/api/specs/routers.yaml:423]: sure, so the path will be `/projects/{project_id}/router-versions` and the query param will be `ensembler_id` and the `status`, when there are no query, should i just return an array of `router versions` from 1 project ?

File: sdk/turing/batch/job.py (hunk 2 above)
Comment (deadlycoconuts) [reply sdk/turing/batch/job.py:148]: Awesome! Thanks for confirming!

File: api/turing/api/ensemblers_api.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (deadlycoconuts) [reply api/turing/api/ensemblers_api.go:130]: > If the only thing that updated were the names

So if I understood correctly, does this refer to the mlflow experiment name? In other words, the ensembler name?

[the remark this answers is not in the document]
Comment (zenovore) [reply api/turing/api/ensemblers_api.go:130]: yepp, the name refer to ensembler name, but if the ensembler name changed, the mlflow experiment name wont change

File: api/turing/api/ensemblers_api.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (deadlycoconuts) [reply api/turing/api/ensemblers_api.go]: > yep, that's correct, if the update failed, then it will delete the new mlflow run

Ah okay, my bad, I forgot you needed to use `pyFuncEnsembler.RunID` since `ensembler` does not contain a `RunID` field.

> the reason i dont handle this error were to avoid confusion, e.g user update, then get failed message that this error happens because we couldn't remove the new run

Hmm... that's a valid concern but could we instead just wrap that a message saying that the deletion of the new run failed while trying to clean up a failed update?

[the remark this answers is not in the document]
Comment (zenovore) [reply api/turing/api/ensemblers_api.go]: > Hmm... that's a valid concern but could we instead just wrap that a message saying that the deletion of the new run failed while trying to clean up a failed update?

Ahh okay, i'll change the handling, so it'll give a error message saying update failed because cleanup process failed

File: api/turing/api/router_versions_api.go [hunk 7]
```diff
@@ -177,6 +180,24 @@ func (c RouterVersionsController) DeployRouterVersion(
 		"version":   int(routerVersion.Version),
 	})
 }
+func (c RouterVersionsController) ListRouterVersionsByEnsembler(
```
Comment (deadlycoconuts) [inline api/turing/api/router_versions_api.go:183]: Since we're changing the name of the API endpoint, can we also change the name of the controller method? :) Thanks!

File: api/turing/api/router_versions_api.go [hunk 8]
```diff
@@ -177,6 +180,24 @@ func (c RouterVersionsController) DeployRouterVersion(
...
+	_ *http.Request,
+	vars RequestVars,
+	_ interface{},
+) *Response {
+	options := service.RouterVersionByEnsemblerListOptions{}
```
Comment (deadlycoconuts) [inline api/turing/api/router_versions_api.go:188]: Same comment about this struct; since the new API endpoint name is more generic, the name of this struct can also be more generic too :)

File: api/turing/service/router_version_service.go [hunk 9]
```diff
@@ -31,6 +37,9 @@ type RouterVersionsService interface {
 	FindLatestVersionByRouterID(routerID models.ID) (*models.RouterVersion, error)
 	// Delete Deletes the given RouterVersion from the db. This method deletes all child objects (enricher, ensembler).
 	Delete(routerVersion *models.RouterVersion) error
+	// FindRouterVersionsByEnsembler Finds router version with status parameter
+	// that use ensembler with id = ensemblerID
+	FindRouterVersionsByEnsembler(options RouterVersionByEnsemblerListOptions) ([]*models.RouterVersion, error)
```
Comment (deadlycoconuts) [inline api/turing/service/router_version_service.go:42]: Same comment here about changing the service method name to make it more consistent with that of the new endpoint name (i.e. we don't need to specify the 'filtering by ensembler' functionality in the service method name).

File: sdk/turing/router/router.py [hunk 10]
```diff
@@ -134,6 +134,21 @@ def get(cls, router_id: int) -> "Router":
             turing.active_session.get_router(router_id=router_id)
         )
 
+    @classmethod
+    def get_router_version_for_ensembler(
```
Comment (deadlycoconuts) [inline sdk/turing/router/router.py:138]: This method can also be renamed with the endpoint name change.

File: sdk/turing/session.py [hunk 11]
```diff
@@ -354,3 +370,19 @@ def get_router_events(self, router_id: int) -> turing.generated.models.RouterEve
         ).projects_project_id_routers_router_id_events_get(
             project_id=self.active_project.id, router_id=router_id
         )
+
+    @require_active_project
+    def get_router_version_for_ensembler(
```
Comment (deadlycoconuts) [inline sdk/turing/session.py:375]: This method name can be changed too :)

File: api/turing/api/ensemblers_api.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (deadlycoconuts) [reply api/turing/api/ensemblers_api.go:130]: > We are not checking for the name change (mlflow experiment name) anywhere in this code, right? 

I think what Alex means is that even though we don't check the name change, mlflow will do it implicitly when the SDK calls the necessary mlflow API endpoints to update the experiment. CMIIW, if mlflow detects only a name change, the run id wouldn't change, so we'd need to check to prevent the existing run id from being deleted.

File: ui/src/ensembler/components/table/ListEnsemblingJobsForEnsemblerTable.js [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+      let inactiveItems = data.results.filter((item) => item.status === 'failed_submission' || item.status==='failed_building' || item.status==='failed' || item.status==='completed');
+      let activeItems = data.results.filter((item) => item.status !== 'failed_submission' && item.status!=='failed_building' && item.status!=='failed' && item.status!=='completed');
+      setResults({
+        inactiveItems: inactiveItems ,
+        activeItems: activeItems ,
+        totalInactiveCount: inactiveItems.length,
+        totalActiveCount : activeItems.length
+      });
+      if (activeItems.length > 0){
+        setDisablePopup(true)
```
Comment (krithika369) [inline ui/src/ensembler/components/table/ListEnsemblingJobsForEnsemblerTable.js:43]: Do we need to set this from here? Instead, can we just set the results from here and use another `useEffect` hook to enable the popup according to the results?

File: sdk/turing/ensembler.py [hunk 13]
```diff
@@ -279,6 +280,24 @@ def update(
             with the model. This will be passed to turing.ensembler.PyFunc.initialize().
             Example: {"config" : "config/staging.yaml"}
         """
+
+        relatedRouterVer = turing.Router.get_router_version_for_ensembler(
+            ensembler_id=self._id, status=[RouterStatus.PENDING]
+        )
+        if len(relatedRouterVer) > 0:
```
[the remark this answers is not in the document]
Comment (deadlycoconuts) [reply sdk/turing/ensembler.py:287]: > Do we need to duplicate these checks here? Won't the API do them anyway and return the same / similar error?

On this, I think the motivation was to prevent the creation of a new mlflow run in the first place (done in the SDK) if we shouldn't be updating the ensembler in the first place (if it's used by a router or a ensembling job). If we were to let the API handle the error, the new mlflow run would have already been created (though the API would rightfully delete the new run if it finds that the ensembler is being used a router or ensembling job). On hindsight, I think maybe we don't need so much redundancy in the code; we can just keep the code simple and rely on the API server to do the heavy lifting and checks.

File: api/turing/api/ensemblers_api.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (zenovore) [reply api/turing/api/ensemblers_api.go:130]: > I think what Alex means is that even though we don't check the name change, mlflow will do it implicitly when the SDK calls the necessary mlflow API endpoints to update the experiment. CMIIW, if mlflow detects only a name change, the run id wouldn't change, so we'd need to check to prevent the existing run id from being deleted.

Yes, That's Correct, if the only thing that changed is the name, mlflow won't create a new run, meaning that the runID wont change, so we only need to check if the RunID is changing

File: sdk/turing/ensembler.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (zenovore) [reply sdk/turing/ensembler.py:287]: > Do we need to duplicate these checks here? Won't the API do them anyway and return the same / similar error?

I move the check from the api to the SDK, the update API doesn't have any more checking, as Zi Yi saying, the main purpose of putting the check in the SDK was to prevent the creation of a new mlflow run in the first place.

What do you guys think?, it is ok to let the SDK create the mlflow creation, since the API also clean it up (meaning that the check will be moved to the API instead of SDK). My idea at first was just to prevent more action (creation and deletion of the mlflow run)

File: api/turing/api/ensembling_job_api.go [hunk 14]
```diff
@@ -138,9 +138,16 @@ func (c EnsemblingJobController) DeleteEnsemblingJob(
...
-	err = c.EnsemblingJobService.MarkEnsemblingJobForTermination(ensemblingJob)
-	if err != nil {
-		return InternalServerError("unable to delete ensembling job", err.Error())
+	if ensemblingJob.Status == models.JobFailed {
+		err = c.EnsemblingJobService.Delete(ensemblingJob)
+		if err != nil {
+			return InternalServerError("unable to delete ensembling job", err.Error())
+		}
+	} else {
+		err = c.EnsemblingJobService.MarkEnsemblingJobForTermination(ensemblingJob)
```
[the remark this answers is not in the document]
Comment (zenovore) [reply api/turing/api/ensembling_job_api.go:147]: The worker will delete job with terminating state
https://github.com/caraml-dev/turing/blob/58f66852ccb6861cda4a4854d6e658ebd72cf8d3/api/turing/batch/ensembling/runner.go#L166-L168

File: api/turing/api/ensemblers_api.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (krithika369) [reply api/turing/api/ensemblers_api.go:130]: I see.. Ok. @zenovore, could you add this explanation to the inline comments too? It makes it clearer. Thanks.

File: sdk/turing/ensembler.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (krithika369) [reply sdk/turing/ensembler.py:287]: Ok, thanks for the clarification. Makes sense to prevent creation via the SDK's checks. Let's keep it this way.

File: api/turing/api/ensemblers_api.go [hunk 15]
```diff
@@ -109,18 +111,172 @@ func (c EnsemblersController) UpdateEnsembler(
...
+	activeOption := service.RouterVersionByEnsemblerListOptions{
+		ProjectID:   options.ProjectID,
+		EnsemblerID: options.EnsemblerID,
+		Statuses:    routerVersionStatusActive,
+	}
+	activeRouter, err := c.RouterVersionsService.FindRouterVersionsByEnsembler(activeOption)
+	if err != nil {
+		return InternalServerError("failed to delete an ensembler", err.Error())
+	}
+	if len(activeRouter) >= 1 {
```
[the remark this answers is not in the document]
Comment (zenovore) [reply api/turing/api/ensemblers_api.go:187]: Just to clarify, this new API will be used only in frontend right? since the API of router version deletion already handle this, so we dont need to check in the backend again.

File: ui/src/components/confirmation_modal/ConfirmationModal.scss [hunk 16]
```diff
@@ -0,0 +1,3 @@
+.euiModalBody__overflow {
+    -webkit-mask-image: none !important;
```
[the remark this answers is not in the document]
Comment (zenovore) [reply ui/src/components/confirmation_modal/ConfirmationModal.scss:2]: Sometimes the content of the popup will become red, i dont know why, previously the content were covered by the mask image fill, for now i cant rpelicate that problem again, but if i dont add the css, it will become like this

File: docs/how-to/delete-an-ensembler/delete-an-ensembler-inactive.md [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+
+Deleting the ensembler will also delete any inactive entity related to the ensembler
+
+Navigate to Ensemblers page
+
+Click on the delete button 
+
+![](../../.gitbook/assets/ensembler_page.png)
+
+The Modal will show related entity that is going to be deleted following the deletion of the ensembler. the router version and the ensembling job shown on the popup will get deleted since both are inactive.
```
Comment (krithika369) [inline docs/how-to/delete-an-ensembler/delete-an-ensembler-inactive.md:13]: Nit: `entity that is` -> `entities that are` (in plural, since there can be more than one)

File: docs/how-to/delete-an-ensembler/delete-an-ensembler.md [hunk 18]
```diff
@@ -0,0 +1,18 @@
+# Deleting an Ensembler without any related entity
+
+An Ensembler only could be deleted if there are no active ensembling jobs / router version that use the ensembler. This page show the deletion of ensembler **without any related entity** (router version and ensembling job)
```
Comment (krithika369) [inline docs/how-to/delete-an-ensembler/delete-an-ensembler.md:3]: Nit on some typos:
* `ensembling jobs / router version` -> `ensembling jobs / router versions`
* `This page show` -> `This page shows`

File: sdk/turing/ensembler.py [hunk 19]
```diff
@@ -279,6 +280,31 @@ def update(
...
+        # First we need to check if there are active job / router using the ensembler
+        # This check is done in the SDK to prevent the creation of new mlflow run
+
+        # check any active router version
+        relatedRouterVer = turing.Router.list_router_versions_with_filter(
+            ensembler_id=self._id, status=[RouterStatus.PENDING], is_current=False
+        )
+        if len(relatedRouterVer) > 0:
+            # if there is any active router version, the deletion process are restricted
+            raise ValueError("There is pending router version using this ensembler")
```
Comment (krithika369) [inline sdk/turing/ensembler.py:293]: Can we also include an actionable message along with the error, like:
```
There is a pending router version using this ensembler. Please wait for the router version to be deployed or undeploy it, before updating the ensembler.
```

File: sdk/turing/ensembler.py [hunk 20]
```diff
@@ -279,6 +280,31 @@ def update(
...
+            ensembler_id=self._id,
+            status=[
+                EnsemblingJobStatus.BUILDING,
+                EnsemblingJobStatus.PENDING,
+                EnsemblingJobStatus.RUNNING,
+            ],
+        )
+        if len(relatedJob) > 0:
+            # if there is any active ensembling jobs, the deletion process are restricted
+            raise ValueError("There is pending ensembling job using this ensembler")
```
Comment (krithika369) [inline sdk/turing/ensembler.py:306]: Same comment as for the routers - will be useful to have some instruction to the user on what needs to be done to overcome this error.

File: ui/src/ensembler/components/modal/DeleteEnsemblerModal.js [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+        color: "success",
+        iconType: "check",
+      });
+      onSuccess();
+      closeModal();
+    }
+  }, [isLoaded, error, ensembler, onSuccess, closeModal]);
+
+  const updateStatus = (newStatus) => {
+    if ((canDeleteEnsembler && newStatus) || (!canDeleteEnsembler && !newStatus)) {
```
Comment (krithika369) [inline ui/src/ensembler/components/modal/DeleteEnsemblerModal.js:46]: * Per your logic, `canDeleteEnsembler` can be set to `false` even when `newStatus ` is `true`.  Is this intentional?
* For consistency, can this function be named to match the `canDeleteEnsembler` state? Also, I think the check can be more concise:
```javascript
const updateCanDeleteEnsembler = (newCanDeleteEnsembler) => {
  if (canDeleteEnsembler !== newCanDeleteEnsembler) {
    setCanDeleteEnsembler(false);
  }
}
```

Now, I have a more basic question - why do we even need this function? Can we directly pass the `setCanDeleteEnsembler` down? IIRC the change will only be triggered by `useState` if old and new values are not the same.

File: ui/src/ensembler/list/ListEnsemblersTable.js [hunk 22]
```diff
@@ -46,6 +52,11 @@ export const ListEnsemblersTable = ({
 
   const onTableChange = ({ page = {} }) => onPaginationChange(page);
 
+  const handleDeleteEnsembler = (ensembler) => {
```
Comment (krithika369) [inline ui/src/ensembler/list/ListEnsemblersTable.js:55]: Nit: Can we call this function `onDeleteEnsembler`? That's the convention we typically use.

File: ui/src/ensembler/components/table/ListEnsemblingJobsForEnsemblerTable.js [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+  const [results, setResults] = useState({ inactiveItems: [], activeItems:[], totalInactiveCount: 0, totalActiveCount:0 });
+
+  const {
+    appConfig: {
+      tables: { defaultTextSize },
+    },
+  } = useConfig();
+
+  const [{ data, isLoaded, error }] = useTuringApi(
+    `/projects/${projectID}/jobs?ensembler_id=${ensemblerID}`,
```
Comment (krithika369) [inline ui/src/ensembler/components/table/ListEnsemblingJobsForEnsemblerTable.js:28]: Since this API is paginated, we need to call it in a loop, one page at a time, right?

File: ui/src/ensembler/components/table/ListEnsemblingJobsForEnsemblerTable.js [hunk 24]
```diff
@@ -0,0 +1,120 @@
...
+    []
+  )
+
+  useEffect(() => {
+    if (isLoaded && !error) {
+      let inactiveItems = data.results.filter((item) => item.status === 'failed_submission' || item.status==='failed_building' || item.status==='failed' || item.status==='completed');
```
Comment (krithika369) [inline ui/src/ensembler/components/table/ListEnsemblingJobsForEnsemblerTable.js:34]: May be clearer to do something similar to an `IN` and `NOT IN` comparison for these `inactiveItems` and `activeItems` check. See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes

You've used this function in some other parts of the implementation, you can use it here too.

File: ui/src/jobs/components/modal/DeleteJobModal.js [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+    },
+    {},
+    false
+  );
+
+  useEffect(() => {
+    if (!isEmpty(job)  && isLoaded && !error) {
+      addToast({
+        id: `submit-success-terminate-${job.id}`,
+        title: `Ensembling Jobs ${job.name} for ensembler ${job.ensembler_id} has been terminated!`,
```
Comment (krithika369) [inline ui/src/jobs/components/modal/DeleteJobModal.js:32]: Nit: `Ensembling Jobs` -> `Ensembling Job` since we are terminating only one ?

File: ui/src/jobs/components/modal/DeleteJobModal.js [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+        iconType: "check",
+      });
+      onSuccess();
+      closeModal();
+    }
+  }, [isLoaded, error, job, onSuccess, closeModal]);
+
+  return (
+    <ConfirmationModal
+      title={["failed", "failed_submission", "failed_building", "completed"].includes(job.status) ? 'Delete Ensembling Jobs' : 'Terminate Ensembling Jobs'}
```
Comment (krithika369) [inline ui/src/jobs/components/modal/DeleteJobModal.js:43]: Since this same check (on the status being `IN` or `NOT IN` this list of statuses) is repeated in multiple places, can we refactor this into a common function in `ui/src/services/job/JobStatus.js` and use that ? Like:
```javascript
const isActiveJobStatus = function(jobStatus) {
    return ["failed", "failed_submission", "failed_building", "completed"].includes(jobStatus);
}
```

**Review by krithika369 — COMMENTED** [review]
@zenovore, thanks for addressing all the previous review comments. The PR largely looks good now, but I left some new comments (most of them are minor). Please have another look. Thanks!

File: api/turing/api/ensemblers_api.go [hunk 27]
```diff
@@ -109,18 +111,221 @@ func (c EnsemblersController) UpdateEnsembler(
...
+		})
+
+	if err != nil {
+		return NotFound("ensembler not found", err.Error())
+	}
+
+	// First we need to check if there are active job / router using the ensembler
+	// If such ensembler exist, the deletion process are restricted
+	// Check if there are any deployed / pending router version using the ensembler
+	httpStatus, err := c.checkActiveRouterVersion(options)
```
[the remark this answers is not in the document]
Comment (zenovore) [reply api/turing/api/ensemblers_api.go:178]: i dont think we need to put that to check again, since deleting the router version that is the current version already prohibited byt the deletion itself (There are already existing check in the delete router version (we cant delete current router version)

File: ui/src/ensembler/components/modal/DeleteEnsemblerModal.js (hunk 21 above)
Comment (zenovore) [reply ui/src/ensembler/components/modal/DeleteEnsemblerModal.js]: the api call was inside a component and the main reason for the updateStatus function is to see if one of the api call (either ensembling job or router version, if one of them have condition that make the deletion of ensembler prohibited, then the other one won't show it's entity. for example : the ensembler have 2 active router version and 3 inactive ensembling jobs, since the router version block the deletion, the popup won't show the 3 inactive ensembling jobs.

But i understand of some confusion that i create, i have refactor the updateStatus to become 2 state, for each the router version and the ensembling jobs, please take a look. Thanks!

File: ui/src/ensembler/components/table/ListEnsemblingJobsForEnsemblerTable.js (hunk 23 above)
Comment (zenovore) [reply ui/src/ensembler/components/table/ListEnsemblingJobsForEnsemblerTable.js:28]: Since the dialog doesnt implement pagination, it will have infinite scroll, so we dont need to set pagination for the api call

File: api/turing/api/ensemblers_api.go [hunk 28]
```diff
@@ -109,18 +111,221 @@ func (c EnsemblersController) UpdateEnsembler(
...
+	if len(activeRouter) >= 1 {
+		return http.StatusBadRequest, fmt.Errorf("there are active router version using this ensembler")
+	}
+	return http.StatusOK, nil
+}
+func (c EnsemblersController) checkActiveEnsemblingJob(options EnsemblersPathOptions) (int, error) {
+	ensemblingJobActiveOption := service.EnsemblingJobListOptions{
+		EnsemblerID: options.EnsemblerID,
+		Statuses: []models.Status{
+			models.JobPending,
```
[the remark this answers is not in the document]
Comment (zenovore) [reply api/turing/api/ensemblers_api.go:259]: `JobTerminating` is considered active status, i already add a message on the modal to help the user to wait until the terminating process is done

File: ui/src/ensembler/components/table/ListEnsemblingJobsForEnsemblerTable.js (hunk 23 above)
Comment (deadlycoconuts) [reply ui/src/ensembler/components/table/ListEnsemblingJobsForEnsemblerTable.js:28]: > Since the dialog doesnt implement pagination, it will have infinite scroll, so we dont need to set pagination for the api call

I think Krithika's referring to a concern whereby the API returns the list of ensemblers in a paginated response by default (e.g. with the default page length set as 10, and only the first page gets returned by default), in which case what we normally do in the UI was to use some nifty trick to issue repeated API calls to get the entire list of _something_, if the response is paginated by default.

That's the case for XP, where the API response to get a list of experiments, for example is paginated by default if we are using neither field selectors (e.g. `id`, `name`, etc. to get specific fields only in the response) nor pagination options.

However, it seems like the way Turing's list ensemblers API endpoint works differently - there's not setting of any default page length if it's not specified anywhere (i.e. the entire list gets returned in the first page, see [this](https://github.com/caraml-dev/turing/blob/main/api/turing/service/pagination.go#L38)). So in this case I don't think there's a need to loop through the API responses in this case.

**Review by deadlycoconuts — APPROVED** [review]
Thanks a lot for the mammoth amount of work on this PR! It looks good on a whole, just left a couple of questions to confirm my understanding of certain parts but otherwise, I think this PR's good to go :) There are some minor parts that can probably be simplified but at this point of time, I think it's already pretty awesome as things are :)

File: api/turing/api/ensemblers_api.go [hunk 29]
```diff
@@ -109,18 +111,248 @@ func (c EnsemblersController) UpdateEnsembler(
...
+		},
+	}
+	inactiveEnsemblingJobs, err := c.EnsemblingJobService.List(ensemblingJobInactiveOption)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+	if inactiveEnsemblingJobs.Paging.Total > 0 {
+		if results, ok := inactiveEnsemblingJobs.Results.([]*models.EnsemblingJob); ok {
+			for _, ensemblingJob := range results {
+				if ensemblingJob.Status == models.JobFailed {
```
[the remark this answers is not in the document]
Comment (zenovore) [reply api/turing/api/ensemblers_api.go:315]: wait i have a reason to do this, but let me check first to confirm the issue, i think it was mainly because the failed one doesn't need to be cleaned up again from the spark, but other error need cleanup action on the spark

File: docs/how-to/delete-an-ensembler/README.md [hunk 30]
```diff
@@ -0,0 +1,7 @@
+# Deleting Ensembler
+
+An ensembler that is used by router version or ensembling jobs can be deleted. An ensembler only can be deleted if it is not in active use in routers or ensembling jobs. Deleting an ensembler will result in the purging of the related router version, ensembling job, and ensembler metadata from turing database. This action is **irreversible**.
```
[the remark this answers is not in the document]
Comment (zenovore) [reply docs/how-to/delete-an-ensembler/README.md:3]: hmm at first i just want to tell that the ensembler could be deleted, but yeah it kinda create a confusion, let me change the sentences so it will be clearer

File: sdk/turing/ensembler.py [hunk 31]
```diff
@@ -279,6 +280,35 @@ def update(
             with the model. This will be passed to turing.ensembler.PyFunc.initialize().
             Example: {"config" : "config/staging.yaml"}
         """
+
+        # First we need to check if there are active job / router using the ensembler
+        # This check is done in the SDK to prevent the creation of new mlflow run
+
+        # check any active router version
+        relatedRouterVer = turing.Router.list_router_versions_with_filter(
+            ensembler_id=self._id, status=[RouterStatus.PENDING], is_current=False
```
[the remark this answers is not in the document]
Comment (zenovore) [reply sdk/turing/ensembler.py:289]: if we want to update the ensembler, we dont need to check deployed router version since it was already build with the old ensembler, if the user want to update the ensembler used by the router version, they just need to redeploy it to apply the new ensembler changes. wdyt?

[the remark this answers is not in the document]
Comment (deadlycoconuts) [reply sdk/turing/ensembler.py:289]: Good point. I was initially worried about what might happen to a router using the current version of the ensembler, should that version be deleted. But you're right, it shouldn't ever be an issue :)

File: api/turing/api/ensemblers_api.go (hunk 28 above)
[the remark this answers is not in the document]
Comment (zenovore) [reply api/turing/api/ensemblers_api.go:261]: oop sorry, this was defined on the frontend, i will also update from the api side

File: api/turing/api/ensemblers_api.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (zenovore) [reply api/turing/api/ensemblers_api.go:317]: im not sure about the jobDeletion Spark Application Status, but after seeing the implementation of the code 
```go
func (c *ensemblingController) Delete(namespace string, ensemblingJob *models.EnsemblingJob) error {
	sa, err := c.clusterController.GetSparkApplication(context.Background(), namespace, ensemblingJob.Name)
	if err != nil {
		c.cleanup(ensemblingJob.Name, namespace)
		// Not found, we do not consider this as an error because its just no further
		// action required on this part.
		return nil
	}
	c.cleanup(ensemblingJob.Name, namespace)
	return c.clusterController.DeleteSparkApplication(context.Background(), namespace, sa.Name)
}
```

if we mark all to terminating, its ok since if the job doesnt exist, it wont show any error, and it will have the cleanup if the job have a spark application / image building job on kaniko

So if we just do MarkEnsemblingJobForTermination for all job deletion handling, its ok, since they will handle if the job not found from kaniko/spark. If theres no data on kaniko/spark they will just to try delete the data from database.

File: api/turing/api/ensemblers_api.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (deadlycoconuts) [reply api/turing/api/ensemblers_api.go]: My bad, I forgot that it's possible to have router versions that are not active but are 'current'. Thanks a lot for clarifying this @zenovore!

**Comment by deadlycoconuts** [comment]
LGTM! 🚀 Feel free to merge this once the pipeline passes! Thanks a lot for working on all the comments!
